Add fulfillments tracking code

tpaksu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Fixed WOOPLUG-5197

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

Implements comprehensive telemetry for the Fulfillments feature to measure adoption, usage patterns, friction points, and underutilized features (WOOPLUG-5197).

#### Tracked Events

**1. Core Funnel (Adoption)**
- `fulfillment_modal_opened` - When a merchant opens the fulfillment editor drawer (frontend JS)
- `fulfillment_created` - When a new fulfillment is saved (REST controller)
- `fulfillment_updated` - When an existing fulfillment is modified (REST controller)
- `fulfillment_deleted` - When a fulfillment is deleted (REST controller)

**2. Tracking Information (Usage Patterns)**
- `fulfillment_tracking_added` - When tracking info (number, provider, URL) is attached to a fulfillment. Determines entry method (`ui_auto_lookup`, `ui_manual_select`, `ui_manual_custom`, or `api`) from the shipping option meta and request source.
- `fulfillment_tracking_lookup_attempted` - When the tracking number auto-lookup is attempted, including whether a URL was generated

**3. Efficiency / Power-User**
- `fulfillment_bulk_action_used` - When bulk fulfill/unfulfill is used from the orders list
- `fulfillment_filter_used` - When orders are filtered by fulfillment status

**4. Customer Communication**
- `fulfillment_notification_sent` - When a fulfillment email is queued to the customer
- `fulfillment_email_template_customized` - When a merchant saves fulfillment email template settings (hooked into `woocommerce_update_options_email_{id}`)

**5. Friction / Errors**
- `fulfillment_validation_error` - When a create/update/delete action fails validation

#### Key Implementation Details

- `FulfillmentsTracker` class centralizes all tracking methods with comprehensive docblocks
- `determine_tracking_entry_method()` helper maps `_shipping_option` meta to standardized entry method values
- `maybe_track_tracking_added()` on the REST controller fires `fulfillment_tracking_added` after create/update when tracking info is present
- Email template tracking hooks into existing WooCommerce email settings save actions
- Frontend event renamed from `woocommerce_fulfillment_drawer_open` to `fulfillment_modal_opened` to align with the spec

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes #60040 WOOPLUG-5197.

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|        |       |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://developer.woocommerce.com/docs/contribution/testing/writing-high-quality-testing-instructions/), include your detailed testing instructions:

1. Enable the fulfillments feature flag (`woocommerce_feature_fulfillments_enabled` = `yes`)
2. Enable tracking (`woocommerce_allow_tracking` = `yes`)
3. Open the browser developer tools Network tab and filter for `t.gif` (Tracks pixel requests)
4. Navigate to **WooCommerce > Orders** and click on a fulfillment trigger button to open the drawer. Verify the `fulfillment_modal_opened` event fires with `source` and `order_id` properties.
5. Create a new fulfillment with tracking info (enter a tracking number, use auto-lookup). Verify:
   - `fulfillment_created` event fires with `source`, `initial_status`, `fulfillment_type`, `item_count`, `total_quantity`
   - `fulfillment_tracking_added` event fires with `entry_method=ui_auto_lookup`, `provider_name`, `is_custom_provider`
   - `fulfillment_tracking_lookup_attempted` fires with `lookup_status`, `provider_identified`, `url_generated`
6. Update the fulfillment, verify `fulfillment_updated` event fires
7. Delete the fulfillment, verify `fulfillment_deleted` event fires
8. Go to **WooCommerce > Settings > Emails**, find a fulfillment email template, save its settings. Verify `fulfillment_email_template_customized` fires.
9. Use the bulk "Mark as fulfilled" action from the orders list. Verify `fulfillment_bulk_action_used` fires.
10. Filter orders by fulfillment status. Verify `fulfillment_filter_used` fires.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

- PHPUnit tests added and passing for `FulfillmentsTrackerTest` (21 tests), `FulfillmentsManagerTest` email hook test (1 test), `OrderFulfillmentsRestControllerTest` tracking integration tests (5 tests)
- PHP linting passes
- PHPStan checked (no new errors introduced)

### Milestone

<!-- DO NOT remove or modify this section (other than to check the box). -->

<!-- milestone-target-selection -->
- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**
<!-- /milestone-target-selection -->

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Add telemetry for fulfillment feature adoption and usage tracking

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

tpaksu committed:
Add fulfillments tracking code

Commit:
Add changefile(s) from automation for the following project(s): woocommerce, woocommerce/client/admin

tpaksu committed:
Merge branch 'trunk' into add/60040-telemetry-for-fulfillments

tpaksu committed:
Merge branch 'trunk' into add/60040-telemetry-for-fulfillments

tpaksu committed:
Merge branch 'trunk' into add/60040-telemetry-for-fulfillments

tpaksu committed:
Merge remote-tracking branch 'origin/trunk' into add/60040-telemetry-for-fulfillments

# Conflicts:
#	plugins/woocommerce/src/Admin/Features/Fulfillments/Fulfillment.php
#	plugins/woocommerce/src/Admin/Features/Fulfillments/FulfillmentsTracker.php
#	plugins/woocommerce/src/Admin/Features/Fulfillments/OrderFulfillmentsRestController.php

tpaksu committed:
Wire up remaining fulfillment telemetry events and add tests

- Add track_fulfillment_modal_opened method and rename JS event to
  fulfillment_modal_opened to match the spec
- Wire up track_fulfillment_tracking_added in REST controller create
  and update flows via maybe_track_tracking_added helper
- Add url_generated param to tracking lookup attempt event
- Hook fulfillment email template save actions to track customization
- Add determine_tracking_entry_method helper for mapping shipping
  option meta to entry method values
- Add comprehensive docblock comments to all tracker methods
- Add FulfillmentsTrackerTest (21 tests) covering entry method logic
  and method signatures
- Add REST controller tests for tracking wiring and request source
- Add FulfillmentsManager test for email template tracking hooks

woocommercebot committed:
Add changefile(s) from automation for the following project(s): woocommerce, woocommerce/client/admin

tpaksu committed:
Fix PHPStan and lint errors in fulfillment tracking code

- Cast $order->get_item_count() to int for track_fulfillment_creation
- Default nullable get_status() to 'unfulfilled' for tracker calls
- Simplify catch-block action to 'update' since $next_state is not
  available when an exception is thrown before assignment
- Add @phpstan-ignore for WP_REST_Request generics on new method

tpaksu committed:
Merge branch 'trunk' into add/60040-telemetry-for-fulfillments

tpaksu committed:
Fix TypeError: cast Exception::getCode() to string for tracker calls

Exception::getCode() returns int, but track_fulfillment_validation_error
expects string. Cast to (string) in all three catch blocks.

tpaksu committed:
Fix array alignment in FulfillmentsTrackerTest data provider

tpaksu committed:
Merge branch 'trunk' into add/60040-telemetry-for-fulfillments

tpaksu committed:
Address PR review comments

- Fix legacy order edit page detection by also checking for 'post'
  query param in isOrderDetailsPage (index.tsx)
- Guard get_item_count() against malformed _items metadata by
  validating each entry is an array with a numeric qty (Fulfillment.php)
- Move filter tracking call after validation and add tracking to the
  legacy orders list query path (FulfillmentsRenderer.php)
- Fix $item_count docblock to reflect it is total quantity, not
  unique line item count (FulfillmentsTracker.php)
- Fix test fixture meta key from _shipment_provider to
  _shipping_provider to match the Fulfillment model accessor

tpaksu committed:
Merge branch 'trunk' into add/60040-telemetry-for-fulfillments

tpaksu committed:
Fix merge conflict style errors

tpaksu committed:
Use typed accessors instead of raw get_meta for fulfillment properties

- Replace get_meta('_shipment_provider') with get_shipping_provider()
  in FulfillmentsRenderer, fixing the meta key mismatch
- Revert unnecessary defensive checks in get_item_count() since items
  are validated before being written
- Update FulfillmentsRendererTest to use set_tracking_number(),
  set_tracking_url(), and set_shipping_provider() typed setters

ayushpahwa commented on plugins/woocommerce/src/Admin/Features/Fulfillments/Fulfillment.php:
```diff
@@ -264,6 +264,23 @@ public function set_items( array $items ): void {
+	 * Get the item count for the fulfillment.
+	 *
+	 * This method calculates the total quantity of items in the fulfillment.
+	 *
```
Given this method is public, we should add a @since annotation here.

tpaksu replied:
Fixed in 4d3ff3a69e.

Added `@since 10.7.0` annotation to the public `get_item_count()` method.

ayushpahwa commented on plugins/woocommerce/src/Admin/Features/Fulfillments/OrderFulfillmentsRestController.php:
```diff
@@ -369,22 +422,28 @@ public function update_fulfillment( WP_REST_Request $request ): WP_REST_Response
 			);
-		} catch ( \Throwable $e ) {
+		} catch ( \Exception $e ) {
+			$action = 'update';
```
nit: I guess this variable and the similar one above isn't used more than once.

tpaksu replied:
Fixed in 4d3ff3a69e.

Inlined the `$action` variable directly as the string `'update'` in both catch blocks since it was only used once.

ayushpahwa commented on plugins/woocommerce/src/Admin/Features/Fulfillments/OrderFulfillmentsRestController.php:
```diff
@@ -261,22 +274,36 @@ public function create_fulfillment( WP_REST_Request $request ) {
+				$this->check_request_source( $request ),
+				$fulfillment->get_is_fulfilled() ? 'fulfilled' : 'draft',
+				$fulfillment->get_item_count() === (int) $order->get_item_count() ? 'full' : 'partial',
+				$fulfillment->get_item_count(),
+				(int) $order->get_item_count(),
+				$notify_customer
+			);
+
+			// Track if tracking information was added with this new fulfillment.
+			$this->maybe_track_tracking_added( $fulfillment, $request );
+
 			if ( $fulfillment->get_is_fulfilled() && $notify_customer ) {
 				/**
 				 * Trigger the fulfillment created notification on creating a fulfilled fulfillment.
 				 *
 				 * @since 10.1.0
 				 */
 				do_action( 'woocommerce_fulfillment_created_notification', $order_id, $fulfillment, wc_get_order( $order_id ) );
+				FulfillmentsTracker::track_fulfillment_notification_sent( 'fulfillment_created', $fulfillment->get_id(), $order_id );
 			}
 		} catch ( ApiException $ex ) {
+			FulfillmentsTracker::track_fulfillment_validation_error( 'create', $ex->getErrorCode(), $this->check_request_source( $request ) );
 			return $this->prepare_error_response(
 				$ex->getErrorCode(),
 				$ex->getMessage(),
 				WP_Http::BAD_REQUEST
 			);
-
-		} catch ( \Throwable $e ) {
+		} catch ( \Exception $e ) {
```
Wouldn't exception be narrower than throwable? Why the change?

tpaksu replied:
Good question. Yes, `\Exception` is narrower than `\Throwable`. The change is intentional: `\Throwable` also catches `\Error` types (TypeError, ParseError, etc.), which represent programming bugs rather than expected runtime errors. In a REST API context, those bugs should propagate to the WordPress error handler rather than being silently returned as 400 responses with telemetry tracking. The `ApiException` catch already handles known validation errors, and `\Exception` covers other runtime exceptions, which is the appropriate scope for this error tracking.

ayushpahwa commented on plugins/woocommerce/src/Admin/Features/Fulfillments/OrderFulfillmentsRestController.php:
```diff
@@ -415,13 +474,21 @@ public function delete_fulfillment( WP_REST_Request $request ) {
 			$fulfillment = new Fulfillment( $fulfillment_id );
 			$this->validate_fulfillment( $fulfillment, $fulfillment_id, $order_id );
 			$fulfillment->delete();
+			FulfillmentsTracker::track_fulfillment_deletion(
+				$this->check_request_source( $request ),
+				$fulfillment_id,
+				$fulfillment->get_status() ?? 'unfulfilled',
+				$notify_customer
+			);
```
I feel this should happen before the delete or the status field should be computed before the delete is done. WDYT?

tpaksu replied:
Fixed in 4d3ff3a69e.

Good catch. The status is now captured before `delete()` is called, so the tracking always records the correct pre-deletion status.

ayushpahwa reviewed (commented):
The code looks good. I tried to follow the testing instructions but only saw the modal opened event and not the other events like created or track attempted. I am guessing this is due to these events being backend only. I tested by adding debug log to the record_event function to test this but I am assuming there might be a better way. If so please do share. Also, shared a couple of questions below, please check out. Thanks

![Image](https://github.com/user-attachments/assets/2678551a-2c5a-4359-a5ec-9aa76f32a836)

tpaksu commented:
<!-- linear:replyTo:3980045371 -->

I think the easiest way to test this would be creating a JN site with this branch, and monitor the tracks events on [https://mc.a8c.com/tracks/live/?eventname=&user=tpaksu&useragent=](<https://mc.a8c.com/tracks/live/?eventname=&user=tpaksu&useragent=>) filtered with the user id. It'll show the events in the last 2 hours and may have some delay before the event shows up.

tpaksu committed:
Address code review feedback for fulfillment telemetry

- Add @since 10.7.0 annotation to public get_item_count() method
- Inline $action variable in update_fulfillment catch blocks
- Capture fulfillment status before delete() call for accurate tracking

tpaksu committed:
Gate fulfillment filter tracking on explicit filter submission

Only track fulfillment_filter_used when the user explicitly clicks
the Filter button (filter_action present), not on pagination, sorting,
or page refresh where fulfillment_status persists in the URL.

tpaksu commented on plugins/woocommerce/src/Admin/Features/Fulfillments/FulfillmentsRenderer.php:
```diff
@@ -469,6 +470,7 @@ public function filter_orders_list_table_query( $args ) {
 
 			// Ensure the fulfillment status is one of the allowed values.
 			if ( FulfillmentUtils::is_valid_order_fulfillment_status( $fulfillment_status ) ) {
+				FulfillmentsTracker::track_fulfillment_filter_used( 'fulfillment_status', $fulfillment_status );
```
Fixed in 1d8778eb63.

Good catch. The tracking now only fires when `filter_action` is present in the GET params, which indicates the user explicitly clicked the Filter button. Pagination, sorting, and page refreshes carry `fulfillment_status` in the URL but not `filter_action`, so those no longer emit duplicate events. Applied the fix to both `filter_orders_list_table_query()` (HPOS) and `filter_legacy_orders_list_query()` (legacy).

ayushpahwa committed:
Merge branch 'trunk' into add/60040-telemetry-for-fulfillments

ayushpahwa commented on plugins/woocommerce/src/Admin/Features/Fulfillments/FulfillmentsTracker.php:
```diff
@@ -0,0 +1,359 @@
+		WC_Tracks::record_event(
+			'fulfillment_created',
+			array(
+				'source'            => $source,
```
What do you think about adding the provider name here?

tpaksu replied:
I believe this won't be needed as the following tracking number added event will record it. If we need to track if the fulfillment was initially created with a tracking number and provider added, we can add it then, but I believe that won't be useful much.

ayushpahwa commented on plugins/woocommerce/src/Admin/Features/Fulfillments/FulfillmentsTracker.php:
```diff
@@ -0,0 +1,359 @@
+			array(
+				'fulfillment_id'     => $fulfillment_id,
+				'entry_method'       => $entry_method,
+				'provider_name'      => $provider_name,
```
The provider name came up blank for me even though it was identified as UPS. 
```
{
"fulfillment_id":"18",
"entry_method":"ui_auto_lookup",
"provider_name":"",
"is_custom_provider":"0"
}
```

tpaksu replied:
That's a great catch.

tpaksu replied:
Some parts of the code were using `_shipping_provider` as the meta key, and some parts were using `_shipment_provider`.

ayushpahwa commented on plugins/woocommerce/src/Admin/Features/Fulfillments/FulfillmentsTracker.php:
```diff
@@ -0,0 +1,359 @@
+				'source'            => $source,
+				'fulfillment_id'    => $fulfillment_id,
+				'original_status'   => $original_status,
+				'changed_fields'    => $changed_fields,
```
I am not able to see the changed fields in the tracked events. Maybe this is due to save being called before computing the diff?

![Image](https://github.com/user-attachments/assets/d1a09eab-2c63-41d8-be0d-66da87a3bfd8)

tpaksu replied:
Fixed. There were two issues. First, the Fulfillment setters wrote directly to $this->data instead of using $this->set_prop(), so get_changes() never tracked core data changes (status, etc.). Second, meta-based fields (tracking_number, items, shipping_provider) use update_meta_data() which doesn't feed into get_changes() at all.

To fix those, I added a protected $data array with defaults and migrated all core setters/getters to use set_prop()/get_prop(). Now status, is_fulfilled, entity_type, etc. are properly tracked. Also, I added a snapshot mechanism. The data store calls `$data->snapshot_meta()` after reading from DB, which captures all meta key/value pairs to memory. Then `get_changes()` is overridden to compare current meta against the snapshot, returning changes under a `meta_data` key. `apply_changes()` refreshes the snapshot.

Now `get_changes()` returns a clean structure like:

```php
[
    'status'       => 'fulfilled',
    'is_fulfilled' => true,
    'meta_data'    => [
         '_tracking_number' => '1Z999...',
         '_shipping_provider' => 'ups',
    ],
]
```

The data store's old `snapshot_tracked_state()`, `compute_changed_props()`, and `TRACKED_META_KEYS` were removed since `get_changes()` now handles everything.

ayushpahwa commented on plugins/woocommerce/src/Admin/Features/Fulfillments/FulfillmentsTracker.php:
```diff
@@ -0,0 +1,359 @@
+ * 5. Friction / Errors:
+ *    - fulfillment_validation_error : A create/update/delete action fails validation.      (REST controller)
+ *
+ * @since 10.1.0
```
This looks incorrect given this class is new. Also we should be adding the since tag to this class' functions too.

tpaksu replied:
Thanks for pointing that out, fixed all.

ayushpahwa reviewed (commented):
Thanks for making the changes. Most of the events look good. I had an issue with 2, shared the comments below.

tpaksu committed:
Merge branch 'trunk' into add/60040-telemetry-for-fulfillments

tpaksu committed:
Address telemetry review feedback and fix fulfillment change tracking

- Fix blank provider name for auto-lookup by adding FulfillmentUtils::resolve_provider_name()
- Fix empty changed_fields by migrating Fulfillment setters to set_prop/get_prop and
  adding meta snapshot-based change detection via get_changes() override
- Fix @since tags on FulfillmentsTracker (10.1.0 -> 10.7.0) and add to all public methods
- Replace data store snapshot_tracked_state/compute_changed_props with Fulfillment::get_changes()
- Simplify determine_tracking_entry_method to 2 params (ui_manual replaces custom/select split)
- Gate fulfillment_tracking_added on actual tracking meta changes during updates
- Update woocommerce_fulfillment_after_update hook to pass $changes and $previous_status
- Fix test flakiness: by-ref parameter errors, stale DB table flag, feature flag pollution
- Add tests for get_changes(), resolve_provider_name, and updated hook signature

tpaksu committed:
Align shipment provider meta key to _shipment_provider across backend

The frontend uses _shipment_provider but the backend Fulfillment model
was reading _shipping_provider, causing the provider to always be empty
for UI-created fulfillments. Rename the meta key, model methods, and all
references to use _shipment_provider consistently.

tpaksu committed:
Add return type to Fulfillment::apply_changes() for PHPStan

ayushpahwa reviewed (approved):
Changes look good and the testing instructions now work as expected. Thanks for working on this. Good to 🚢 

(Worth noting that 10.7.0 is on code freeze so this pr might need to be cherry picked or change the tags to the next version.)

tpaksu committed:
Merge branch 'trunk' into add/60040-telemetry-for-fulfillments
